### PR TITLE
feat(vpto): add pto.fdiv_hp - A5 SIMT scalar correctly-rounded FP32 division (#1117)

### LINKED ISSUE #1117: [Bug] [PTODSL] [阻塞] FP32 标量 lowering 未经 fastmath 请求收缩 FMA，且 divf 存在 1 ULP 差异
### Component

Python bindings (python/)

### Description

SIMT FP32 标量算术存在两类精度语义差异：

1. PTODSL 生成独立的 `arith.mulf` 和 `arith.addf`，且没有设置
   `fastmath`/`contract`，但最终运行结果与融合的 `fmaf` 完全一致。
2. SIMT 标量 `arith.divf` 与 PyTorch NPU FP32 除法存在约 1 ULP 差异，
   当前也没有类似 Tile `tdiv` 的 high-precision 标量接口可供选择。

### Reproduction (minimal)

```markdown
import ctypes

import torch
import torch_npu
from ptodsl import pto


N = 256

torch.npu.config.allow_internal_format = False
torch_npu.npu.set_compile_mode(jit_compile=False)
torch.npu.set_device("npu:0")


@pto.simt(max_threads=N)
def div_simt(
    lhs: pto.ptr(pto.f32, "gm"),
    rhs: pto.ptr(pto.f32, "gm"),
    dst: pto.ptr(pto.f32, "gm"),
):
    tid = pto.get_tid_x()
    pto.stg(pto.ldg(lhs, tid) / pto.ldg(rhs, tid), dst, tid)


@pto.jit(
    name="simt_div_probe",
    kernel_kind="vector",
    target="a5",
    backend="vpto",
    mode="explicit",
)
def div_kernel(
    lhs: pto.ptr(pto.f32, "gm"),
    rhs: pto.ptr(pto.f32, "gm"),
    dst: pto.ptr(pto.f32, "gm"),
):
    div_simt[N, 1, 1](lhs, rhs, dst)


@pto.simt(max_threads=N)
def muladd_simt(
    a: pto.ptr(pto.f32, "gm"),
    b: pto.ptr(pto.f32, "gm"),
    c: pto.ptr(pto.f32, "gm"),
    dst: pto.ptr(pto.f32, "gm"),
):
    tid = pto.get_tid_x()
    pto.stg(
        pto.ldg(c, tid) + pto.ldg(a, tid) * pto.ldg(b, tid),
        dst,
        tid,
    )


@pto.jit(
    name="simt_muladd_probe",
    kernel_kind="vector",
    target="a5",
    backend="vpto",
    mode="explicit",
)
def muladd_kernel(
    a: pto.ptr(pto.f32, "gm"),
    b: pto.ptr(pto.f32, "gm"),
    c: pto.ptr(pto.f32, "gm"),
    dst: pto.ptr(pto.f32, "gm"),
):
    muladd_simt[N, 1, 1](a, b, c, dst)


def mismatch_count(actual, expected):
    return int(
        (actual.view(torch.int32) != expected.view(torch.int32)).sum()
    )


stream = torch.npu.current_stream()._as_parameter_

# divf
torch.manual_seed(20260803)

lhs_host = torch.rand(N, dtype=torch.float32) * 2
rhs_host = torch.rand(N, dtype=torch.float32) * 128 + 0.25

lhs = lhs_host.to("npu:0")
rhs = rhs_host.to("npu:0")
div_output = torch.empty_like(lhs)

compiled_div = div_kernel.compile()
compiled_div[1, stream](
    lhs.data_ptr(),
    rhs.data_ptr(),
    div_output.data_ptr(),
)
torch.npu.synchronize()

actual_div = div_output.cpu()
expected_div = (lhs / rhs).cpu()

print(
    "div mismatch:",
    mismatch_count(actual_div, expected_div),
    "/",
    N,
)
print(
    "div max abs:",
    float((actual_div - expected_div).abs().max()),
)

# mulf + addf
torch.manual_seed(20260803)

a_host = torch.randn(N, dtype=torch.float32)
b_host = torch.randn(N, dtype=torch.float32)
c_host = torch.randn(N, dtype=torch.float32)

a = a_host.to("npu:0")
b = b_host.to("npu:0")
c = c_host.to("npu:0")
muladd_output = torch.empty_like(a)

compiled_muladd = muladd_kernel.compile()
compiled_muladd[1, stream](
    a.data_ptr(),
    b.data_ptr(),
    c.data_ptr(),
    muladd_output.data_ptr(),
)
torch.npu.synchronize()

actual_muladd = muladd_output.cpu()
separate_muladd = (c + a * b).cpu()

libm = ctypes.CDLL("libm.so.6")
fmaf = libm.fmaf
fmaf.argtypes = [
    ctypes.c_float,
    ctypes.c_float,
    ctypes.c_float,
]
fmaf.restype = ctypes.c_float

fused_muladd = torch.tensor(
    [
        fmaf(float(x), float(y), float(z))
        for x, y, z in zip(a_host, b_host, c_host)
    ],
    dtype=torch.float32,
)

print(
    "mul+add vs separate mismatch:",
    mismatch_count(actual_muladd, separate_muladd),
    "/",
    N,
)
print(
    "mul+add vs fmaf mismatch:",
    mismatch_count(actual_muladd, fused_muladd),
    "/",
    N,
)

# 查看 PTODSL 生成的 MLIR
mlir_text = muladd_kernel.compile().mlir_text()
for line in mlir_text.splitlines():
    if "mulf" in line or "addf" in line or "fastmath" in line:
        print(line)
```

### Expected behavior

1. 当 MLIR 中只有独立的 arith.mulf 和 arith.addf，且没有fastmath<contract> 时，后端应保留两次独立舍入，不应自动收缩为 FMA。
2. 如果 PTOAS 设计上允许默认收缩，应提供显式控制开关，并明确记录数值语义。
3. SIMT FP32 标量除法应提供正确舍入或 high-precision 模式，类似 Tile tdiv(..., div_precision=...) 的能力。
4. 应增加 A5 VPTO 端到端运行测试，逐位验证：
  - 独立 mulf + addf
  - 显式 pto.fma
  - 标量 arith.divf
  - high-precision 标量除法

### Actual behavior / error logs

```text
div mismatch: 22 / 256
div max abs: 2.9802322387695312e-08

mul+add vs separate mismatch: 61 / 256
mul+add vs fmaf mismatch: 0 / 256


%7 = arith.mulf %4, %6 : f32
%8 = arith.addf %2, %7 : f32
```

### Git commit

7fd64fe21a403ae85288d8bcdeef34cb333f985

### Host platform

None

### Target Ascend arch (if relevant)

None

### PTOAS build level (if relevant)

None

### PR DESCRIPTION
**Pull request by HecreReed**
## Summary

Implements the remaining item of **Issue #1117**: a scalar **FP32 correctly-rounded division** (round-to-nearest-even) for the A5 SIMT pipeline, exposed as a dedicated PTODSL/MLIR operation instead of relying on the target's native `fdiv` precision (which can differ by ~1 ULP from PyTorch NPU).

Constraints honored:

- **A5 + scalar f32** in this first version. The enforcement of "A5 only" is at the document/API level (the op lives in the VPTO SIMT dialect and the expansion emits target-neutral `arith` ops); a target gate may be added later if non-A5 VPTO targets need distinguishing.
- Plain Python `/` in PTODSL is untouched — it still emits `arith.divf`.
- Does **not** reuse the Tile `precisionType` machinery.
- Does **not** call `llvm.hivm.div.full.f32` / `.rn.f32` (they fail with `Cannot select: intrinsic` on A5 Bisheng, per the issue investigation).

## API / IR

```python
result = pto.fdiv_hp(lhs, rhs)
```

```mlir
%r = pto.fdiv_hp %lhs, %rhs : f32, f32 -> f32
```

## Implementation (`ExpandFDivHpPattern`, `vpto-expand-wrapper-ops`)

`pto.fdiv_hp` is expanded into a pure `arith` integer sequence (no i64 division, no math dialect dependency):

1. Both operands are bit-cast to `i32` and split into sign / 8-bit exponent / 23-bit fraction.
2. **Special inputs** (NaN, Inf, zero dividend, zero divisor) select the raw `arith.divf` result at the end, preserving the backend's native special-value behavior bit-for-bit; the divisor is kept non-zero on the integer path so division-by-zero is never executed.
3. Finite non-zero operands become exact integers `value = M * 2^E`:
   - normal: `M = 0x800000 | fraction`, `E = exponent - 150`;
   - subnormal: `M = fraction` normalized to bit 23 with a compare-shift **CLZ** (checks at 16/24/28/30/31, shifts 16/8/4/2/1) while `E` is adjusted (`math.ctlz` is not used because its A5 SIMT lowering could not be verified; the compare-shift variant only needs shifts/compares/selects that the SIMT pipeline already lowers).
4. The exact quotient `raw = floor(Ma * 2^26 / Mb)` (with `Ma/Mb` normalized into `[1, 2)`) and its remainder come from an unrolled base-2^7 long division. That is one mathematical division, implemented as **9 base-2^7 digit steps (9 `arith.divui`)**; every intermediate fits in `i32`, so no i64 division is introduced.
5. **RNE rounding** with guard/round/sticky (all flag arithmetic in `i32`; predicates are lifted with `sel(c, 1, 0)`): `sig = raw >> 3`, `inc = guard && (round || sticky || sig&1)`; a carry to `1<<24` shifts `sig` and bumps the exponent.
6. The same `raw` also feeds the **subnormal** result path through exact right shifts (`q = (Ma << (e+152)) / Mb` ≡ `raw >> (26 - shift)`, with the remainder sticky `(raw & (2^s - 1)) != 0 || rem != 0` plus the quotient LSB). Shift amounts are clamped to `[0, 31]` so the (selected-away) degenerate cases never produce LLVM-poison shifts.
7. Overflow → ±Inf, underflow → signed zero, and subnormal rounding past `1<<23` → minimum normal `2^-126` are handled when reassembling the result bits.

The algorithm was validated with a **bit-exact Python mirror** of the expansion vs an **independent exact-rational oracle** (`fractions.Fraction` + RNE, itself cross-checked against the host FPU): **0 mismatches over ~600k randomized + exhaustive-significand + edge cases** (normals, 1/7, 7/3, min/max normal, subnormals, ties/midpoints, underflow, overflow, ±0, ±Inf, NaN).

## Tests

- `ptodsl/tests/test_fdiv_hp_algorithm.py` — pure-stdlib software model vs independent oracle (no NPU needed; also registered as ctest `ptodsl_fdiv_hp_algorithm`). **CI: PASSED.**
- `ptodsl/tests/test_fdiv_hp_compile.py` — PTODSL frontend emits the `pto.fdiv_hp` op; plain `/` still emits `arith.divf`. **CI: PASSED.**
- `test/lit/vpto/fdiv_hp_expand_vpto_llvm.pto` — `vpto-expand-wrapper-ops` removes `pto.fdiv_hp` and produces the `arith` expansion; the expanded sequence must survive to `convert-func-to-llvm` (`llvm.udiv`). **CI: PASSED** (build-and-test).
- `ptodsl/tests/e2e/test_fdiv_hp.py` — A5 board validation (`@pytest.mark.require_npu`): compares **raw uint32 bit patterns** (never allclose) against the independent oracle. Includes the **fixed issue #1117 mismatch inputs** (torch seed 20260803), the full coverage list above, and a special-input check that `pto.fdiv_hp` matches a plain-`divf` kernel bit for bit. Inputs are padded to the 1024-thread grid so no thread runs out of bounds. **Pending A5 hardware.**

## Review fixes

Round 1:

- **Blocking type bug**: several flag combines mixed `i1` and `i32` operands (`arith.ori`/andi require identical operand types), which would make the pass output fail verification on any real use. All flag arithmetic now stays in `i32` via a `flag(i1) -> i32` lift; comparisons remain `i1` only where `arith.select` consumes them.
- **Poison shifts**: subnormal-path shift amounts are now clamped to `[0, 31]` (`safeShift`), eliminating LLVM-poison shifts in the degenerate (selected-away) branch.
- **Lit test**: made function-name independent (the pre-pass pipeline can rename the pto.simt_entry function) and checks only robust points (after-pass expansion + LLVM lowering); the frontend-emission assertion lives in the compile test.

Round 2:

- **Board test OOB (P1)**: the kernels launch a fixed 1024-thread grid with unconditional ldg/stg, but the corpus has only 549 elements; threads 549..1023 would access past the tensors. Inputs are now padded to the grid and only the first n_valid entries are compared.
- **Subnormal sweep direction (P2)**: (1.0, 2^-k) computes 2^k (large normals/overflow), not a subnormal quotient; the tiny value is now the DIVIDEND so the quotient 2^-k exercises the subnormal output path (including the k=150 half-ulp tie). Fixed in both tests.
- **Special-value sign loss (P3)**: the software model used x*y>0 for the result sign, which collapses to false for +/-0 products (1/0 returned -Inf, 0/+Inf returned -0). The model now uses copysign-based sign XOR and the special-value table asserts full bit patterns (NaN payloads stay exponent-only checks).

## Notes / known limitations

- First version: A5 + scalar f32 only; f16/bf16/packed and non-A5 targets not yet supported (documented in `docs/isa/micro-isa/17-simt.md`).
- Performance: the expansion is ~200 arith ops (including the 9-digit unrolled 32-bit division), so `pto.fdiv_hp` is much slower than `arith.divf` — use it only where correctly rounded scalar FP32 division matters.
- The Tile high-precision division (`lib/TileOps/a5/div_hp.py`) was deliberately **not** copied: its three-candidate search steps by ±1.0 ULP and its dividend-subnormal detection only against the max subnormal are not suitable for a correctly-rounded scalar implementation.
- CI status: **all checks green**, including vpto-sim-validation. (One earlier vpto-sim run failed on a transient infrastructure error — `shallow.lock` in the shared LLVM cache on the simulator runner — and passed on re-run.)

Fixes #1117

### CODE REVIEW

**Review by Zhendong404 — CHANGES_REQUESTED** [review]
建议先从接口和 lowering 架构上调整，再合入本 PR。

理想方案是增加统一的 scalar `pto.fdiv` semantic op，并在 DSL 上提供类似：

```python
pto.fdiv(lhs, rhs, precision="default")
```

其中 `precision` 作为 op attribute（沿用 `tdiv` 的 `precisionType` 模式）：

- `default`：lower 到 native `arith.divf`；
- `high_precision`：由 A5 SoftLib 提供高精度实现；
- SoftLib 不可用时，才使用受控的软件 fallback，不能静默退回普通 `divf`。

这样可以把精度语义保留到 target lowering 阶段，由后端选择 intrinsic、SoftLib 或 fallback，也避免后续继续增加 `fdiv_hp`、`fdiv_exact` 等多个 scalar op。另需明确 `high_precision` 是“精度改善”还是 IEEE correctly-rounded；如果是后者，建议使用更明确的 rounding/accuracy 语义，因为现有 Div754 三候选算法本身没有严格正确舍入的误差界证明。

当前主线还没有通用 scalar `pto.fdiv`：普通 DSL `/` 直接生成 `arith.divf`，PTO dialect 中只有本 PR 新增的 `pto.fdiv_hp`，而 `tdiv/tdivs` 是 tile/DPS op。直接在本 PR 中实现独立的 `fdiv_hp`，会把语义接口、A5 限制和软件展开耦合在一起，也不便于后续 SoftLib 接入。

因此建议等待 PR1189 合入、主线具备 scalar `pto.fdiv` 基础后，再将本 PR 适配到该接口，并把当前长整数展开保留为 SoftLib/fallback 实现。
